src/ophyd_async/core/utils.py

```diff
@@ -11,46 +24,75 @@
+            Mapping of device name to error or subdevice with errors.
+            Alternatively a string with the signal error.
+        """
+
+        self._errors = errors
+
+    def format_error_string(self, errors: ErrorText, indent="") -> str:
+        string = ""
+        if isinstance(errors, str):
+            return f"{errors}\n"
+        elif not isinstance(errors, dict):
+            raise RuntimeError(
+                f"Unknown error type `{type(errors)}` " "expected `str` or `dict`"
+            )
+
+        for name, value in errors.items():
+            string += indent + f"{name}:"
+            if isinstance(value, str):
+                string += " " + f"{value}\n"  # On the same line as the name
+            elif isinstance(value, dict):
+                string += "\n" + self.format_error_string(
+                    value, indent=(indent + self._indent_width)
+                )
+            else:
+                raise RuntimeError(f"`{type(value)}` not a string or a dict")
+        return string
 
     def __str__(self) -> str:
-        return "\n".join(self.lines)
+        return self.format_error_string(self._errors, indent="")
 
 
 async def wait_for_connection(**coros: Awaitable[None]):
-    """Call many underlying signals, accumulating `NotConnected` exceptions
+    """Call many underlying signals, accumulating exceptions and returning them
 
-    Raises
-    ------
-    `NotConnected` if cancelled
+    Expected kwargs should be a mapping of names to coroutine tasks to execute.
     """
-    ts = {k: asyncio.create_task(c) for (k, c) in coros.items()}  # type: ignore
-    try:
-        done, pending = await asyncio.wait(ts.values())
-    except asyncio.CancelledError:
-        for t in ts.values():
-            t.cancel()
-        lines: List[str] = []
-        for k, t in ts.items():
-            try:
-                await t
-            except NotConnected as e:
-                if len(e.lines) == 1:
-                    lines.append(f"{k}: {e.lines[0]}")
-                else:
-                    lines.append(f"{k}:")
-                    lines += [f"  {line}" for line in e.lines]
-        raise NotConnected(*lines)
-    else:
-        # Wait for everything to foreground the exceptions
-        for f in list(done) + list(pending):
-            await f
+    results = await asyncio.gather(*coros.values(), return_exceptions=True)
+    exceptions = {}
+
+    for name, result in zip(coros, results):
+        if isinstance(result, Exception):
+            if isinstance(result, NotConnected):
+                exceptions[name] = result._errors
+            else:
+                exceptions[name] = f"unexpected exception {type(result).__name__}"
+                logging.exception(
+                    f"device `{name}` raised unexpected "
+                    f"exception {type(result).__name__}",
+                    exc_info=result,
+                )
+
+    if exceptions:
+        raise NotConnected(exceptions)
```

Comment:
I wanted to write this, but that would push the string formatting back down and you would lose the dictionary. It would handle multiline exception output though, so maybe this would be better?

```suggestion
    exceptions: Dict[str, Exception] = {}

    for name, result in zip(coros, results):
        if isinstance(result, Exception):
            exceptions[name] = f"{type(result).__name__}: {result}"
            if not isinstance(result, NotConnected):
                logging.exception(f"{self.name}: device `{name}` raised unexpected exception")

    if exceptions:
        raise NotConnected(exceptions)
```